src/Orckestra.Composer.CompositeC1/Factory/ProductPromotionsFactory.cs

```diff
@@ -0,0 +1,89 @@
+﻿using Orckestra.Composer.CompositeC1.Context;
+using Orckestra.Composer.Factory;
+using Orckestra.Composer.Providers;
+using System;
+using System.Collections.Generic;
+using System.Globalization;
+using System.Linq;
+
+namespace Orckestra.Composer.CompositeC1.Factory
+{
+    public class ProductPromotionsFactory : IProductPromotionsFactory
+    {
+        public IProductTileConfigurationContext ProductTileConfigurationContext { get; }
+        protected ILocalizationProvider LocalizationProvider { get; }
+
+        public ProductPromotionsFactory(
+            ILocalizationProvider localizationProvider,
+            IProductTileConfigurationContext productTileConfigurationContext)
+        {
+            ProductTileConfigurationContext = productTileConfigurationContext ?? throw new ArgumentNullException(nameof(productTileConfigurationContext));
+            LocalizationProvider = localizationProvider ?? throw new ArgumentNullException(nameof(localizationProvider));
+        }
+
+        public virtual Dictionary<string, string> BuildProductBadgeValues(string keys, string lookupDisplayNames)
+        {
+            var productBadgesNames = lookupDisplayNames != null ? lookupDisplayNames.Split(',').ToList() : new List<string>();
+            var productBadgeKeys = keys != null ? keys.Split('|').ToList() : new List<string>();
+
+            if (!productBadgeKeys.Any())
+            {
+                return null;
+            }
+
+            var productBadgeValues = new Dictionary<string, string>();
+
+            for (var i = 0; i < productBadgeKeys.Count; i++)
+            {
+                if (!productBadgeValues.ContainsValue(productBadgeKeys[i]))
+                    productBadgeValues.Add(productBadgeKeys[i], productBadgesNames[i]);
```

Comment:
we don't check that productBadgesNames have the same count as productBadgeKeys, so IndexOutOfRangeException  can happen

Reply:
I added unit tests to validate all cases and did changes